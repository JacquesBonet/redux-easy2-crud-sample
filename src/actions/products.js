import PRODUCT from "../constants/products";
import {actions} from "redux-easy2-crud"


export const addProduct = (aProduct) => (actions.create(PRODUCT, aProduct));
export const readProducts = () => (actions.read(PRODUCT));
export const updateProduct = (product) => (actions.update(product));
export const deleteProduct = (product) => (actions.del(product));
