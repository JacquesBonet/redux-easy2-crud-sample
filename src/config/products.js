export const productInfos = {
  FR1: {
    name: 'Fruit Tea',
    prices: [{
      count: -1,
      value: 3.11
    }]
  },
  AP1: {
    name: 'Apple',
    free: ['FR1'],
    prices: [{
      count: 2,
      value: 5.00
    },
    {
      count: -1,
      value: 4.50
    }]
  },
  CF1: {
    name: 'Coffee',
    free: ['FR1'],
    prices: [{
      count: -1,
      value: 11.23
    }]
  }
};


/**
 * Calculate the number of products of type product.productCode
 *
 * @param products       {Array}   Array of products
 * @param code           {String}  The product code to  search products
 * @param position      {int}      The product position on the basket
 * @returns {*}                    return the product number matching productCode code
 */
const getNbProductsWithSameCode = (products, code, position) => {
  return products.reduce((total, product, idx) => {
    if (idx > position) return total;
    if (product.productCode === code) return total + 1;
    return total;
  }, 0)
};

export const getProductCodes = () => (Object.keys(productInfos).sort());

export const getProductName = (productOrCode) => {
  let productCode = productOrCode.productCode ? productOrCode.productCode : productOrCode;
  return productInfos[productCode] ? productInfos[productCode].name : '';
};

/**
 * Calculate the product price
 *
 * @param products      {Array}    Array of products
 * @param product       {Object}   Product to calculate the price
 * @param position      {int}      The product position on the basket
 * @returns {int}                  return the price
 */
export const getProductValue = (products, position = 0) => {
  let product = products[position];
  if (product.hasOwnProperty('price'))  return product.price;

  let nbProducts = getNbProductsWithSameCode( products, product.productCode, position)
  let prices = productInfos[product.productCode].prices;

  for( var price of prices) {
    if (nbProducts <= price.count) return price.value;

    if (price.count === -1) return price.value;
  }
};

/**
 * Return the free products discount associated to a product
 *
 * @param productCode  {Array}    the free product discount. 0, if no product discount
 */
export const hasFreeProductDiscount = (productCode) => {
  const product = productInfos[productCode];

  if (product) return product.free

  return 0;
};


