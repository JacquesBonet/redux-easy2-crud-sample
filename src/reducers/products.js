import PRODUCTS from "../constants/products";
import {crudReducer}  from 'redux-easy2-crud';

const initialState = [];


export default function products(state = initialState, action) {
  return crudReducer(state, PRODUCTS, action);
}
