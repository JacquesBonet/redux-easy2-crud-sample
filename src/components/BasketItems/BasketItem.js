import React, {PropTypes} from "react";
import {getProductName, getProductValue} from "../../config/products";
import formatCurrency from "../../formatCurrency";
import "./Component.css";

const BasketItem = ({products, idx, removeHandler}) => {
  let product = products[idx];
  return (
    <div className={product.price === 0 ? 'basketitem__free' : 'basketitem'}>
      <span className="basketitem__name">{getProductName(product)}</span>
      <span
        className={product.price === 0 ? 'basketitem__freevalue' : 'basketitem__value'}>
         {formatCurrency(getProductValue(products, idx))}
       </span>
      {product.price !== 0 ? <button className="basketitem__button" onClick={removeHandler}>x</button> : ''}
    </div>
  );
};

BasketItem.propTypes = {
  products: PropTypes.arrayOf(PropTypes.object).isRequired,
  idx: PropTypes.number.isRequired,
  removeHandler: PropTypes.func.isRequired
};

export default BasketItem;
