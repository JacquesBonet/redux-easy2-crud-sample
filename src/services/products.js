import {getProductCodes, getProductName, getProductValue, hasFreeProductDiscount} from "../config/products";
export {getProductCodes, getProductName, getProductValue, hasFreeProductDiscount};

/**
 * return the total products price in the basket
 *
 * @param productsState         {Array}    the redux state
 * @returns {*}
 */
export const getTotal = (productsState) => {
  return productsState.reduce((total, current, idx) => {
    return total + getProductValue( productsState, idx);
  }, 0);
};


/**
 * return the list of products
 *
 * @param rootState     {Object}    the redux state
 * @returns {Array}
 */
export const getProducts = (rootState) => (rootState.products);


/**
 * id generation
 * @returns {String}
 */
export const id = () => {
  return Math.random().toString(36).substring(7)
}
