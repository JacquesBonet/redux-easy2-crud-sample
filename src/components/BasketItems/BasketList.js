import React, {PropTypes} from "react";
import BasketItem from './BasketItem';
import "./Component.css";

const BasketList = ({products, deleteProduct}) => (
  <div className="basketlist">
    <br/>
    <ul>
      {
        products.map(( product, idx) => (
          <li key={idx}>
            <BasketItem products={products} idx={idx} removeHandler={deleteProduct.bind(this, product)} />
          </li>
        ))
      }
    </ul>
  </div>
);

BasketList.propTypes = {
  products: PropTypes.arrayOf(PropTypes.object).isRequired,
  deleteProduct: PropTypes.func.isRequired
};

export default BasketList;
