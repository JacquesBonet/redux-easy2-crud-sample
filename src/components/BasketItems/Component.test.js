import React from "react";
import {shallow} from "enzyme";
import BasketList from "./BasketList";
import BasketItem from "./BasketItem";
import {getProductName, getProductValue} from "../../config/products";
import formatCurrency from "../../formatCurrency";

function deleteProduct() {}

describe('BasketItemsComponent', function () {
  let component;
  const products = [{productCode: 'AP1'}, {productCode: 'AP1'}, {productCode: 'CF1'}];
  beforeEach(function () {
    component = shallow(<BasketList products={products} deleteProduct={deleteProduct} />);
  });

  products.forEach(function (product, idx) {
    let productCode = product.productCode;
    let productItemComponent;
    let expectedName = getProductName(product);
    let expectedValue = formatCurrency(getProductValue(products, idx));

    beforeEach(function () {
      productItemComponent = component.find('li').at(idx).find(BasketItem).shallow();
    });

    it(`should have title ${expectedName} for ${productCode}`, function () {
      expect(productItemComponent.find('.basketitem__name').html()).toMatchSnapshot();
    });

    it(`should have value ${expectedValue} for ${productCode}`, function () {
      expect(productItemComponent.find('.basketitem__value').html()).toMatchSnapshot();
    });
  });
});
