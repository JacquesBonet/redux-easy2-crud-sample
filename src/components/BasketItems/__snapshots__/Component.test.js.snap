exports[`BasketItemsComponent should have title Apple for AP1 1`] = `"<span class=\"basketitem__name\">Apple</span>"`;

exports[`BasketItemsComponent should have title Apple for AP1 2`] = `"<span class=\"basketitem__name\">Apple</span>"`;

exports[`BasketItemsComponent should have title Coffee for CF1 1`] = `"<span class=\"basketitem__name\">Coffee</span>"`;

exports[`BasketItemsComponent should have value $5.00 for AP1 1`] = `"<span class=\"basketitem__value\">$5.00</span>"`;

exports[`BasketItemsComponent should have value $5.00 for AP1 2`] = `"<span class=\"basketitem__value\">$5.00</span>"`;

exports[`BasketItemsComponent should have value $11.23 for CF1 1`] = `"<span class=\"basketitem__value\">$11.23</span>"`;
